src/gdb_main.c

```diff
@@ -330,102 +329,155 @@ int gdb_main_loop(struct target_controller *tc, bool in_syscall)
 	}
 }
 
-static void
-handle_q_string_reply(const char *str, const char *param)
+static void handle_q_string_reply(const char *str, const char *param)
 {
 	unsigned long addr, len;
 
-	if (sscanf(param, "%08lx,%08lx", &addr, &len) != 2) {
+	if (sscanf(param, "%08lx,%08lx", &addr, &len) != 2)
+  {
 		gdb_putpacketz("E01");
 		return;
 	}
-	if (addr < strlen (str)) {
-		char reply[len+2];
-		reply[0] = 'm';
-		strncpy (reply + 1, &str[addr], len);
-		if(len > strlen(&str[addr]))
-			len = strlen(&str[addr]);
-		gdb_putpacket(reply, len + 1);
-	} else if (addr == strlen (str)) {
-		gdb_putpacketz("l");
-	} else
-		gdb_putpacketz("E01");
+  if (addr > strlen (str))
+  {
+    	gdb_putpacketz("E01");
+      return;
+  }
+  if(addr== strlen (str))
+  {
+    	gdb_putpacketz("l");
+      return;
+  }
+  unsigned long int outputLen=strlen(str)-addr;
+  if(outputLen>len) outputLen=len;
```

Comment:
The basic concept looks good, the formatting, as you noted, needs re-doing here (whitespace around operators, condensed {} style as shown by the original code around this).